lib/ibm/cloud/sdk/vpc/base_vpc.rb

```diff
@@ -0,0 +1,61 @@
+# typed: true
+# frozen_string_literal: true
+
+module IBM
+  module Cloud
+    module SDK
+      # Container that encapsulates the VPC API.
+      class BaseVPC
+        def initialize(parent, endpoint = nil)
+          @endpoint = parent.url(endpoint)
+          @token = parent.token
+          @logger = parent.logger
+        end
+
+        attr_reader :endpoint, :token, :logger
+
+        def adhoc(method: 'get', path: nil, params: {}, payload_type: 'json', payload: {})
+          @token.adhoc(method.to_sym, url(path), metadata(params, payload_type, payload))
+        end
+
+        def get(path: nil, params: {})
+          adhoc(method: 'get', path: path, params: params)
+        end
+
+        def post(payload = {}, path: nil, params: {}, type: 'json')
+          adhoc(method: 'post', path: path, params: params, payload: payload)
+        end
+
+        def put(payload = {}, path: nil, params: {})
+          adhoc(method: 'put', path: path, params: params, payload: payload)
+        end
+
+        def patch(payload = {},  path: nil, params: {})
+          adhoc(method: 'patch', path: path, params: params, payload: payload)
+        end
+
+        def delete(path: nil, params: {})
+          adhoc(method: 'delete', path: path, params: params)
+        end
+
+        def url(path = nil)
+          return endpoint unless path
+
+          "#{endpoint}/#{path}"
+        end
+
+        private
+
+        def metadata(params = {}, payload_type = 'json', payload = {})
+          params ||= {}
```

Comment:
This would cover you if someone passed in `metadata(nil)` but if that's what the concern is you could just leave `params = nil` as the default method parameter and do `params ||= {}` here or `.merge(params || {})` below

Reply:
Still working through this. I changed the http provider and am still figuring out the best way of dealing with the options.